File: templates/wasm-node-nim/src/node.nim
## Flow-Like WASM Node Template (Nim)
##
## A template for creating custom nodes in Nim that compile to WebAssembly
## via Nim → C → Emscripten.
##
## Building:
##   nimble build
##
## The compiled .wasm file will be at: build/node.wasm

import sdk

# ============================================================================
# Node Definition
# ============================================================================

proc buildDefinition(): NodeDefinition =
  var def = initNodeDefinition()
  def.name = "my_custom_node_nim"
  def.friendlyName = "My Custom Node (Nim)"
  def.description = "A template WASM node built with Nim"
  def.category = "Custom/WASM"
  def.addPermission("streaming")

  # Input pins
  def.addPin inputPin("exec", "Execute", "Trigger execution", Exec)
  def.addPin inputPin("input_text", "Input Text", "Text to process", String).withDefault("\"\"")
  def.addPin inputPin("multiplier", "Multiplier", "Number of times to repeat", I64).withDefault("1")

  # Output pins
  def.addPin outputPin("exec_out", "Done", "Execution complete", Exec)
  def.addPin outputPin("output_text", "Output Text", "Processed text", String)
  def.addPin outputPin("char_count", "Character Count", "Number of characters in output", I64)

  def

# ============================================================================
# Node Execution
# ============================================================================

proc handleRun(ctx: var Context): ExecutionResult =
  let inputText = ctx.getString("input_text")
  let multiplier = ctx.getI64("multiplier", 1)

  ctx.debug("Processing: '" & inputText & "' x " & $multiplier)

  var output = ""
  for i in 0 ..< multiplier:
    output.add inputText

  let charCount = output.len

  ctx.streamText("Generated " & $charCount & " characters")

  ctx.setOutput("output_text", jsonString(output))
  ctx.setOutput("char_count", $charCount)

  ctx.success()

# ============================================================================
# WASM Exports
# ============================================================================

proc get_node(): int64 {.exportc.} =
  serializeDefinition(buildDefinition())

proc get_nodes(): int64 {.exportc.} =
  let def = buildDefinition()
  packResult("[" & def.toJson() & "]")

proc run(p: uint32; l: uint32): int64 {.exportc.} =
  var raw = newString(l)
  if l > 0:
    copyMem(addr raw[0], cast[pointer](p), l)
  let input = parseInput(raw)
  var ctx = newContext(input)
  let res = handleRun(ctx)
  serializeResult(res)

File: templates/wasm-node-lua/src/node.lua
-- Flow-Like WASM Node Template (Lua)
--
-- This file contains your custom node logic.
-- Build: mise run build
-- The compiled .wasm file will be at: build/node.wasm

local sdk = require("sdk")

-- ============================================================================
-- Node Definition
-- ============================================================================

function get_node()
    local def = sdk.newNodeDefinition()
    def.name          = "my_custom_node_lua"
    def.friendly_name = "My Custom Node (Lua)"
    def.description   = "A template WASM node built with Lua"
    def.category      = "Custom/WASM"
    sdk.addPermission(def, "streaming")

    -- Input pins
    sdk.addPin(def, sdk.inputExec())
    sdk.addPin(def, sdk.withDefault(
        sdk.inputPin("input_text", "Input Text", "Text to process", sdk.DataType.String),
        '""'
    ))
    sdk.addPin(def, sdk.withDefault(
        sdk.inputPin("multiplier", "Multiplier", "Number of times to repeat", sdk.DataType.I64),
        "1"
    ))

    -- Output pins
    sdk.addPin(def, sdk.outputExec())
    sdk.addPin(def, sdk.outputPin("output_text", "Output Text", "Processed text", sdk.DataType.String))
    sdk.addPin(def, sdk.outputPin("char_count", "Character Count", "Number of characters in output", sdk.DataType.I64))

    return sdk.serializeDefinition(def)
end

function get_nodes()
    return "[" .. get_node() .. "]"
end

-- ============================================================================
-- Node Execution
-- ============================================================================

function run_node(raw_json)
    local input = sdk.parseInput(raw_json)
    local ctx = sdk.newContext(input)

    local inputText = ctx:getString("input_text", "")
    local multiplier = ctx:getI64("multiplier", 1)
    if multiplier < 0 then multiplier = 0 end

    ctx:debug("Processing: '" .. inputText .. "' x " .. tostring(multiplier))

    local parts = {}
    for i = 1, multiplier do
        parts[i] = inputText
    end
    local outputText = table.concat(parts)
    local charCount = #outputText

    ctx:streamText("Generated " .. tostring(charCount) .. " characters")

    ctx:setOutput("output_text", sdk.jsonString(outputText))
    ctx:setOutput("char_count", tostring(charCount))

    local result = ctx:success()
    return sdk.serializeResult(result)
end
